File: backend/app/schemas/schedule.py
from datetime import datetime
from typing import Optional, List
from uuid import UUID
from pydantic import BaseModel, model_validator
from ..models.schedule import ScheduleStatus, SchedulePriority


class _UserSnap(BaseModel):
    id: UUID
    full_name: str
    username: str
    model_config = {"from_attributes": True}


class ScheduleCreate(BaseModel):
    mbr_id: UUID
    planned_batch_size: Optional[float] = None
    batch_unit: Optional[str] = None
    scheduled_start: datetime
    scheduled_end: datetime
    assigned_operator_id: Optional[UUID] = None
    equipment_line: Optional[str] = None
    priority: SchedulePriority = SchedulePriority.MEDIUM
    notes: Optional[str] = None

    @model_validator(mode="after")
    def end_after_start(self):
        if self.scheduled_end <= self.scheduled_start:
            raise ValueError("scheduled_end must be after scheduled_start")
        return self


class ScheduleUpdate(BaseModel):
    scheduled_start: Optional[datetime] = None
    scheduled_end: Optional[datetime] = None
    assigned_operator_id: Optional[UUID] = None
    equipment_line: Optional[str] = None
    priority: Optional[SchedulePriority] = None
    notes: Optional[str] = None
    planned_batch_size: Optional[float] = None
    batch_unit: Optional[str] = None


class ScheduleOut(BaseModel):
    id: UUID
    schedule_number: str
    mbr_id: UUID
    mbr_number: str
    mbr_version: str
    product_name: str
    product_code: str
    planned_batch_size: Optional[float] = None
    batch_unit: Optional[str] = None
    scheduled_start: datetime
    scheduled_end: datetime
    equipment_line: Optional[str] = None
    status: ScheduleStatus
    priority: SchedulePriority
    notes: Optional[str] = None
    ebr_id: Optional[UUID] = None
    converted_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    assigned_operator: Optional[_UserSnap] = None
    converted_by: Optional[_UserSnap] = None
    created_by: _UserSnap

    model_config = {"from_attributes": True}


class ScheduleList(BaseModel):
    items: List[ScheduleOut]
    total: int
    page: int
    size: int


class ConvertPayload(BaseModel):
    batch_number: str
    planned_batch_size: Optional[float] = None
    batch_unit: Optional[str] = None
    notes: Optional[str] = None

File: frontend/src/api/schedule.ts
import client from './client';

export type ScheduleStatus   = 'PLANNED' | 'CONFIRMED' | 'IN_PROGRESS' | 'COMPLETED' | 'CANCELLED';
export type SchedulePriority = 'HIGH' | 'MEDIUM' | 'LOW';

export interface UserSnap { id: string; full_name: string; username: string; }

export interface Schedule {
  id: string;
  schedule_number: string;
  mbr_id: string;
  mbr_number: string;
  mbr_version: string;
  product_name: string;
  product_code: string;
  planned_batch_size: number | null;
  batch_unit: string | null;
  scheduled_start: string;
  scheduled_end: string;
  equipment_line: string | null;
  status: ScheduleStatus;
  priority: SchedulePriority;
  notes: string | null;
  ebr_id: string | null;
  converted_at: string | null;
  created_at: string;
  updated_at: string;
  assigned_operator: UserSnap | null;
  converted_by: UserSnap | null;
  created_by: UserSnap;
}

export interface CalendarEntry {
  id: string;
  schedule_number: string;
  product_name: string;
  product_code: string;
  mbr_number: string;
  status: ScheduleStatus;
  priority: SchedulePriority;
  scheduled_start: string;
  scheduled_end: string;
  equipment_line: string | null;
  ebr_id: string | null;
  assigned_operator: string | null;
  planned_batch_size: number | null;
  batch_unit: string | null;
}

export interface ScheduleList {
  items: Schedule[];
  total: number;
  page: number;
  size: number;
}

export interface ScheduleCreatePayload {
  mbr_id: string;
  planned_batch_size?: number;
  batch_unit?: string;
  scheduled_start: string;
  scheduled_end: string;
  assigned_operator_id?: string;
  equipment_line?: string;
  priority?: SchedulePriority;
  notes?: string;
}

export const scheduleApi = {
  list: (params?: { page?: number; size?: number; status?: string; search?: string }) =>
    client.get<ScheduleList>('/schedule', { params }).then(r => r.data),

  calendar: (start: string, end: string) =>
    client.get<CalendarEntry[]>('/schedule/calendar', { params: { start, end } }).then(r => r.data),

  get: (id: string) =>
    client.get<Schedule>(`/schedule/${id}`).then(r => r.data),

  create: (data: ScheduleCreatePayload) =>
    client.post<Schedule>('/schedule', data).then(r => r.data),

  update: (id: string, data: Partial<ScheduleCreatePayload>) =>
    client.put<Schedule>(`/schedule/${id}`, data).then(r => r.data),

  delete: (id: string) => client.delete(`/schedule/${id}`),

  confirm: (id: string) =>
    client.post<Schedule>(`/schedule/${id}/confirm`).then(r => r.data),

  cancel: (id: string) =>
    client.post<Schedule>(`/schedule/${id}/cancel`).then(r => r.data),

  convert: (id: string, data: { batch_number: string; planned_batch_size?: number; batch_unit?: string; notes?: string }) =>
    client.post<Schedule>(`/schedule/${id}/convert`, data).then(r => r.data),
};
